src/main/scala/mesosphere/marathon/api/akkahttp/Directives.scala

```diff
@@ -25,6 +90,62 @@ object Directives extends AuthDirectives with LeaderDirectives with AkkaDirectiv
     }
   }
 
+  private val marathonApiKeywords = Set("restart", "tasks", "versions")
+
+  //basically it's equivalent to path.takeWhile(s => !marathonApiKeywords.contains(s)) map createPathId
+  private[akkahttp] def findSomethingLikeAppId(remainingPath: Path, extractedAppName: Path = Path.Empty): Option[PathId] = {
+    def generatePathId(p: Path): PathId = {
+      val trimmed = p.toString().reverse.dropWhile(_ == '/').reverse
+      PathId(trimmed)
+    }
+
+    remainingPath match {
+      case Path.Slash(rest) =>
+        findSomethingLikeAppId(rest, extractedAppName ++ Path./)
+      case p @ Path.Segment(segment, rest) =>
+        if (marathonApiKeywords.contains(segment)) {
+          Some(generatePathId(extractedAppName))
+        } else {
+          findSomethingLikeAppId(rest, extractedAppName + segment)
+        }
+      case Path.Empty =>
+        if (extractedAppName.isEmpty) {
+          None
+        } else {
+          Some(generatePathId(extractedAppName))
+        }
+    }
+  }
+
+  /**
+    * Given the current root group, only extract an existing appId
+    *
+    * This is useful because our v2 API has an unfortunate design decision which leads to ambiguity in our URLs, such as:
+    *
+    *   POST /v2/apps/my-group/restart/restart
+    *
+    * The intention here is to restart the app named "my-group/restart"
+    *
+    * This directive will only extract "my-group/restart" from the path, leaving the rest
+    *
+    * In case there is no appId exists, it will do best-efforts attempt to find an invalid appId and generate a rejection.
+    *
+    * Inspired by path matching directive in akka-http.
+    */
+  def extractExistingAppId(rootGroup: RootGroup): Directive1[PathId] = {
```

Comment:
How does this method relate to `ExistingAppPathId` and `RemainingPathId`? It would help a lot to have a few example cases. I really feel this should be a separated PR followed by a PR for the app controller. Trust me, it'd be a much faster review process :).